selfdrive/locationd/laikad.py

```diff
@@ -149,55 +150,83 @@ def read_ephemeris(self, gnss_msg):
-        "correctedMeasurements": meas_msgs
-      }
-      return dat
-    elif self.is_ephemeris(gnss_msg):
+
+    if self.is_ephemeris(gnss_msg):
       self.read_ephemeris(gnss_msg)
+      return None
+
+    if not self.is_good_report(gnss_msg):
+      return None
+
+    #if gnss_msg.which() == 'ionoData':
+    # TODO: add this, Needed to better correct messages offline. First fix ublox_msg.cc to sent them.
+
+    week, tow, new_meas = self.read_report(gnss_msg)
+    if len(new_meas) == 0:
+      return None
+
+    self.gps_week = week
+    t = gnss_mono_time * 1e-9
+    if week > 0:
+      self.got_first_gnss_msg = True
+      latest_msg_t = GPSTime(week, tow)
+      if self.auto_fetch_orbits:
+        self.fetch_orbits(latest_msg_t, block)
+
+    # Filter measurements with unexpected pseudoranges for GPS and GLONASS satellites
+    new_meas = [m for m in new_meas if 1e7 < m.observables['C1C'] < 3e7]
+    processed_measurements = process_measurements(new_meas, self.astro_dog)
+
+    est_pos, prev_est_pos = self.get_est_pos(t, processed_measurements)
+
+    corrected_measurements = []
+    if len(est_pos) > 0:
+      corrected_measurements = correct_measurements(processed_measurements, est_pos, self.astro_dog)
+
+    if gnss_mono_time % 10 == 0:
+      cloudlog.debug(f"Measurements Incoming/Processed/Corrected: {len(new_meas), len(processed_measurements), len(corrected_measurements)}")
+
+    self.update_localizer(est_pos, t, corrected_measurements)
+
+    if est_pos == prev_est_pos or len(prev_est_pos) == 0:
+      # no new measurement
+      return None
+
+    #kf_valid = all(self.kf_valid(t))
+    #ecef_pos = self.gnss_kf.x[GStates.ECEF_POS]
+    #ecef_vel = self.gnss_kf.x[GStates.ECEF_VELOCITY]
+
+    #p = self.gnss_kf.P.diagonal()
+    #pos_std = np.sqrt(p[GStates.ECEF_POS])
+    #vel_std = np.sqrt(p[GStates.ECEF_VELOCITY])
+
+    ecef_vel, vel_res = calc_vel_fix(corrected_measurements, est_pos)
+    ecef_vel = ecef_vel[:3]
+
+    vel_std = np.array([np.abs(np.mean(vel_res))]*3)
+    vel_valid = len(ecef_vel) != 0
+
+    meas_msgs = [create_measurement_msg(m) for m in corrected_measurements]
+    dat = messaging.new_message("gnssMeasurements")
+    measurement_msg = log.LiveLocationKalman.Measurement.new_message
+
+    # process the residual error to get a meaningful accuracy
+    # filter residuals over 400 and calculate the positive mean
+    res = list(filter(lambda x: x < 400, self.last_pos_residual))
+    pos_std = np.array([np.sqrt(np.mean(np.abs(res)))]*3)
+
+    dat.gnssMeasurements = {
+      "gpsWeek": week,
+      "gpsTimeOfWeek": tow,
+      "positionECEF": measurement_msg(value=est_pos, std=pos_std.tolist(), valid=True),
+      "velocityECEF": measurement_msg(value=ecef_vel, std=vel_std.tolist(), valid=vel_valid),
+      # TODO std is incorrectly the dimension of the measurements and not 3D
+      "positionFixECEF": measurement_msg(value=self.last_pos_fix, std=self.last_pos_residual, valid=self.last_pos_fix_t == t),
```

Comment:
Fix this. You can use this function to either get an sensible average value or to compute correctly each iteration
https://github.com/commaai/laika/blob/828612e1b8848ccf70072d5513c0b7977f1707da/laika/raw_gnss.py#L411